pip/pip-420.md

```diff
@@ -0,0 +1,276 @@
+  - Call this method when creating schema
+  - External schema can be initialized when calling this method
+
+- close() **(New addition)**
+  - Called when the producer or consumer is closed.
+  - Allows external schema implementations to release resources, such as schema registry connections or caches.
+
+#### Example Workflow:
+
+- During producer or consumer initialization:
+  The external schema info will be registered to Pulsar schema storage.
+
+- During message send or receive:
+  The `encode` and `decode` methods handle the schema-aware serialization and deserialization using the external schema registry.
+
+#### Schema ID & Schema Version
+
+Unlike Pulsar, which uses **schema version** to identify schemas, many external schema registry systems use **schema ID** as the primary schema identifier.
+
+When integrating with external schema registries:
+- The `schemaVersion` will not indicate the read schema version, it points an external schema.
+- Instead, the external schema implementation can manage schema ID handling internally.
+- The schema ID can be embedded directly into the message payload by the external schema’s `encode` and `decode` methods.
+
+This approach allows external schema systems to fully control schema evolution and versioning without being constrained by Pulsar’s native schema versioning mechanism.
+This may impact some components that rely on schema version to deserialize messages, such as Pulsar Functions and Pulsar SQL;
+they need to be updated to support setting schema properties, using the new schema type and handle external schemas appropriately.
+
+#### Example usage
+
+```java
+public void workWithExternalSchemaRegistry() throws Exception {
+    Map<String, String> srConfig = new HashedMap<>();
+    srConfig.put("schema.registry.url", "http://localhost:8001");
+
+    String topic = "testExternalJsonSchema";
+  
+    Schema<User> schema = KafkaSchemas.JSON(User.class);
+  
+    @Cleanup
+    PulsarClient pulsarClient = PulsarClient.builder()
+            .serviceUrl("pulsar://localhost:" + getBrokerServicePort())
+            .schemaProperties(srConfig)
+            .build();
+  
+    @Cleanup
+    Producer<User> producer = pulsarClient.newProducer(schema)
+            .topic(topic)
+            .create();
+  
+    @Cleanup
+    Consumer<User> consumer = pulsarClient.newConsumer(schema)
+            .topic(topic)
+            .subscriptionName("sub")
+            .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+            .subscribe();
+  
+    for (int i = 0; i < 10; i++) {
+      producer.send(new User("name-" + i, 10 + i));
+    }
+  
+    for (int i = 0; i < 10; i++) {
+      Message<User> message = consumer.receive();
+      consumer.acknowledge(message);
+      assertEquals(message.getValue().getName(), "name-" + i);
+    }
+}
+```
+
+## Public-facing Changes
+
+Introduce a new SchemaType `EXTERNAL` to represent the schema types that work with external schema registry.
```

Comment:
It seems that we need a definition of the interface `external schema store client`, because we need to call it in the Pulsar client.

Reply:
Because the other schema system may integrate register and retrieve schema operations into the serialization and deserialization process, such as Kafka JSON, Avro serializer and deserializer, Pulsar doesn't need to touch that underlying operations; otherwise, we need to rewrite a lot of logic in custom external schemas.